src/core/filesystem.cc

```diff
@@ -497,31 +514,310 @@ Status
+    Aws::Client::ClientConfiguration config(profile_name);
+    client_ = s3::S3Client(config);
+  } else {
+    Aws::Client::ClientConfiguration config("default");
+    client_ = s3::S3Client(config);
+  }
+}
+
+S3FileSystem::~S3FileSystem()
+{
+  Aws::ShutdownAPI(options_);
+}
+
+Status
+S3FileSystem::FileExists(const std::string& path, bool* exists)
+{
+  *exists = false;
+
   std::string bucket, object;
-  ParsePath(path, &bucket, &object);
+  RETURN_IF_ERROR(ParsePath(path, &bucket, &object));
 
-  gcs::ObjectWriteStream stream = client_->WriteObject(bucket, object);
+  // Construct request for object metadata
+  s3::Model::HeadObjectRequest head_request;
+  head_request.SetBucket(bucket.c_str());
+  head_request.SetKey(object.c_str());
 
-  if (!stream) {
+  auto head_object_outcome = client_.HeadObject(head_request);
+  if (head_object_outcome.IsSuccess()) {
+    *exists = true;
+    return Status::Success;
+  }
+
+  // S3 doesn't make objects for directories, so it could still be a directory
+  bool is_dir;
+  RETURN_IF_ERROR(IsDirectory(path, &is_dir));
+  *exists = is_dir;
+
+  return Status::Success;
+}
+
+Status
+S3FileSystem::IsDirectory(const std::string& path, bool* is_dir)
+{
+  *is_dir = false;
+  std::string bucket, object_path;
+  RETURN_IF_ERROR(ParsePath(path, &bucket, &object_path));
+
+  // Check if the bucket exists
+  s3::Model::HeadBucketRequest head_request;
+  head_request.WithBucket(bucket.c_str());
+
+  auto head_bucket_outcome = client_.HeadBucket(head_request);
+  if (!head_bucket_outcome.IsSuccess()) {
     return Status(
         RequestStatusCode::INTERNAL,
-        "failed to open object write stream for " + path);
+        "Could not get MetaData for bucket with name " + bucket);
   }
 
-  stream.Close();
+  // Root case - bucket exists and object path is empty
+  if (object_path.empty()) {
+    *is_dir = true;
+    return Status::Success;
+  }
 
-  google::cloud::StatusOr<gcs::ObjectMetadata> metadata =
-      std::move(stream).metadata();
+  // List the objects in the bucket
+  s3::Model::ListObjectsRequest list_objects_request;
+  list_objects_request.SetBucket(bucket.c_str());
+  list_objects_request.SetPrefix(AppendSlash(object_path).c_str());
+  auto list_objects_outcome = client_.ListObjects(list_objects_request);
 
-  if (!metadata) {
+  if (list_objects_outcome.IsSuccess()) {
+    *is_dir = !list_objects_outcome.GetResult().GetContents().empty();
+  } else {
     return Status(
-        RequestStatusCode::INTERNAL, "Error writing to object at " + path);
+        RequestStatusCode::INTERNAL,
+        "Failed to list objects with prefix " + path);
   }
+  return Status::Success;
+}
 
+Status
+S3FileSystem::FileModificationTime(const std::string& path, int64_t* mtime_ns)
+{
+  // We don't need to worry about the case when this is a GCS directory
+  bool is_dir;
+
+  RETURN_IF_ERROR(IsDirectory(path, &is_dir));
+  if (is_dir) {
+    return Status::Success;
```

Comment:
Is it possible to get the modification time for directory? In model repository manager, I used the modification time of the directory for the case of file deletion ([here](https://github.com/NVIDIA/tensorrt-inference-server/blob/master/src/core/model_repository_manager.cc#L194)). If the way I am doing only make sense on local filesystem, then I will need to use a different approach

Reply:
I can define it as the most recently modified child in the directory, and return that.

Reply:
Seems like these cases should be added to the CI somehow.